storage-narpc/src/main/java/com/ibm/crail/storage/tcp/TcpStorageRequest.java

```diff
@@ -0,0 +1,198 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.ibm.crail.storage.tcp;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import com.ibm.crail.conf.CrailConstants;
+import com.ibm.narpc.NaRPCMessage;
+
+public class TcpStorageRequest implements NaRPCMessage {
+	public static final int CSIZE = 4 + Math.max(WriteRequest.CSIZE, ReadRequest.CSIZE);
+	
+	private int type;
+	private WriteRequest writeRequest;
+	private ReadRequest readRequest;
+	
+	public TcpStorageRequest(){
+		writeRequest = new WriteRequest();
+		readRequest = new ReadRequest();
+	}
+
+	public TcpStorageRequest(WriteRequest writeRequest) {
+		this.writeRequest = writeRequest;
+		this.type = TcpStorageProtocol.REQ_WRITE;
+	}
+
+	public TcpStorageRequest(ReadRequest readRequest) {
+		this.readRequest = readRequest;
+		this.type = TcpStorageProtocol.REQ_READ;
+	}
+
+	public int size() {
+		return CSIZE;
+	}
+	
+	public int type(){
+		return type;
+	}
+
+	@Override
+	public void update(ByteBuffer buffer) throws IOException {
+		type = buffer.getInt();
+		if (type == TcpStorageProtocol.REQ_WRITE){
+			writeRequest.update(buffer);
+		} else if (type == TcpStorageProtocol.REQ_READ){
+			readRequest.update(buffer);
+		}
+	}
+
+	@Override
+	public int write(ByteBuffer buffer) throws IOException {
+		buffer.putInt(type);
+		int written = 4;
+		if (type == TcpStorageProtocol.REQ_WRITE){
+			written += writeRequest.write(buffer);
+		} else if (type == TcpStorageProtocol.REQ_READ){
+			written += readRequest.write(buffer);
+		}
+		return written;
+	}
+	
+	public static class WriteRequest {
+		public static final int CSIZE = 4 + 8 + 4 + 4 + (int) CrailConstants.BLOCK_SIZE;
+		
+		private int key;
+		private long address;
+		private int length;
+		private ByteBuffer data;
+		
+		public WriteRequest(){
+			data = ByteBuffer.allocateDirect((int) CrailConstants.BLOCK_SIZE);
+		}
+		
+		public WriteRequest(int key, long address, int length, ByteBuffer buffer) {
+			this.key = key;
+			this.address = address;
+			this.length = length;
+			this.data = buffer;
+		}
+
+		public long getAddress() {
+			return address;
+		}
+
+		public int length() {
+			return length;
+		}
+		
+		public int getKey(){
+			return key;
+		}
+
+		public ByteBuffer getBuffer() {
+			return data;
+		}
+
+		public int size() {
+			return CSIZE;
+		}
+		
+		public void update(ByteBuffer buffer) throws IOException {
+			key = buffer.getInt();
+			address = buffer.getLong();
+			length = buffer.getInt();
+			int remaining = buffer.getInt();
+			buffer.limit(buffer.position() + remaining);
+			data.clear();
+			data.put(buffer);
+			data.flip();
+		}
+
+		public int write(ByteBuffer buffer) throws IOException {
+			buffer.putInt(key);
+			buffer.putLong(address);
+			buffer.putInt(length);
+			buffer.putInt(data.remaining());
+			int written = 4 + 8 + 4 + 4 + data.remaining(); 
```

Comment:
See above! At least make this one variable e.g. HEADER_SIZE. This is a nightmare if you want to change one of the parameters.